src/Security/EventListener/DenyAccessListener.php

```diff
@@ -56,27 +57,53 @@ public function __construct(ResourceMetadataFactoryInterface $resourceMetadataFa
      * @throws AccessDeniedException
      */
     public function onKernelRequest(GetResponseEvent $event): void
+    {
+        @trigger_error(sprintf('Method "%1$s::onKernelRequest" is deprecated since API Platform 2.4 and will not be available anymore in API Platform 3. Prefer calling "%1$s::onSecurity" instead.', self::class), E_USER_DEPRECATED);
+    }
+
+    public function onSecurity(GetResponseEvent $event): void
+    {
+        $this->checkSecurity($event->getRequest(), 'security');
+    }
+
+    public function onLateSecurity(GetResponseEvent $event): void
     {
         $request = $event->getRequest();
+        $this->checkSecurity($request, 'late_security', [
+            'previous_object' => $request->attributes->get('previous_data'),
+        ]);
+    }
+
+    /**
+     * @throws AccessDeniedException
+     */
+    private function checkSecurity(Request $request, string $attribute, array $variables = []): void
+    {
         if (!$attributes = RequestAttributesExtractor::extractAttributes($request)) {
             return;
         }
 
         $resourceMetadata = $this->resourceMetadataFactory->create($attributes['resource_class']);
 
-        $isGranted = $resourceMetadata->getOperationAttribute($attributes, 'access_control', null, true);
+        $isGranted = $resourceMetadata->getOperationAttribute($attributes, $attribute, null, true);
+        if (null === $isGranted) {
+            // Backward compatibility
+            $isGranted = $resourceMetadata->getOperationAttribute($attributes, 'access_control', null, true);
```

Comment:
It's possible to pass `access_control` in `onKernelRequest`, then you don't need this check anymore.

Reply:
Can't, otherwise I couldn't trigger the deprecation